bin/demo.py

```diff
@@ -2,221 +2,215 @@
+        package_config = options_file.read().replace("JENKINS_NAME", jenkins_name)
+    with open("jenkins_config.json", "w") as options_file:
+        options_file.write(package_config)
+    install_package('jenkins', None, jenkins_name, "jenkins_config.json")
+    os.remove("jenkins_config.json")
+    assert package_installed('jenkins', jenkins_name), log_and_exit('!! package failed to install')
+    log("waiting for Jenkins service to come up at '{}'".format(jenkins_url))
+    end_time = time.time() + 300
+    while time.time() < end_time:
+        if verify_jenkins(jenkins_url):
+            break
+        time.sleep(1)
+
+def verify_jenkins(jenkins_url):
+    try:
+        r = http.get(jenkins_url)
+        if r.status_code == 200 and r.headers['x-jenkins']:
+            log("service is up and running, got Jenkins version '{}'".format(r.headers['x-jenkins']))
+            return True
+    except:
         return False
-    log ("Jenkins is up and running! Got Jenkins version {}".format(r.headers['x-jenkins']))
-    return True
 
-def check_marathon_lb(elb_url):
-    log("Checking to see if Marathon-lb is installed.")
-    try:
-        r = requests.get(elb_url, verify=False)
-        if r.status_code == 503 and not r.text:
-            install_marathon_lb(elb_url)
-        else:
-            log ("Marathon-lb already seems to be running at {}. Not installing Marathon-lb.".format(elb_url))
-    except requests.exceptions.ConnectionError:
-        install_marathon_lb(elb_url)
-
-def install_marathon_lb(elb_url):
-    log ("Couldn't find a Marathon-lb instance running at {}.".format(elb_url))
-    log ("Installing Marathon-lb.")
-    command = "dcos package install --yes marathon-lb"
-    print ("\n> " + command)
-    if call (['dcos', 'package', 'install', '--yes', 'marathon-lb']) != 0:
-        log ("Failed to install Marathon-lb.")
-    else:
-        log("Marathon-lb has been installed!")
+def install_marathon_lb(marathon_lb_url):
+    log("installing marathon-lb")
+    with stdchannel_redirected(sys.stdout, os.devnull):
+        run_dcos_command('marathon app add conf/get_sa.json')
+        end_time = time.time() + 300
+        while time.time() < end_time:
+            if get_marathon_task('saread'):
+                break
+            time.sleep(1)
+        log("retrieving service account JSON")
+        satoken = run_dcos_command("task log --lines=1 saread")[0]
+        run_dcos_command('marathon app remove saread')
+    post_url = "{}secrets/v1/secret/default/marathon-lb".format(dcos_url)
+    headers = {'Content-Type': 'application/json'}
+    data = json.dumps({ 'value' : satoken })
+    r = http.put(post_url, headers=headers, data=data)
+    install_package('marathon-lb', None, None, "conf/marathon-lb.json")
+
+def verify_marathon_lb(marathon_lb_url):
+    with stdchannel_redirected(sys.stderr, os.devnull):
+        try:
+            r = http.get(marathon_lb_url)
+            if r.status_code == 200 and r.text:
```

Comment:
Marathon-lb will only return a 200 if there is an application being served. If there is no application being served, it returns a 503 with text in the body (e.g. `503 Service Unavailable No server is available to handle this request.`). If there is no Marathon-lb, then the Amazon ELB itself returns a 503 - but without any text in the body.

(This check doesn't currently work, you may want to consider using the 503 approach or a more elegant version of it:)
```
➜  cd-demo git:(sunil-demo2) ✗ python3 bin/demo.py pipeline --password=$PASSWORD --branch=sunil-demo2 http://sunil-f4j-publicsl-9adnwfg46s3s-1406902471.us-west-2.elb.amazonaws.com/ http://sunil-f4j-elasticl-1hx6ny1cqwidv-1482094555.us-west-2.elb.amazonaws.com/
[demo]: couldn't find marathon-lb running at 'http://sunil-f4j-publicsl-9adnwfg46s3s-1406902471.us-west-2.elb.amazonaws.com/'
[demo]: installing marathon-lb

》installing package 'marathon-lb'

We recommend a minimum of 0.5 CPUs and 256 MB of RAM available for the Marathon-LB DCOS Service.
Marathon-lb DCOS Service has been successfully installed!
See https://github.com/mesosphere/marathon-lb for documentation.
On branch sunil-demo2
Untracked files:
	conf/cd-demo-app.json
	conf/get_sa.json
	conf/marathon-lb.json
	tmp/

nothing added to commit but untracked files present
[demo]: !! failed to commit updated marathon.json
➜  cd-demo git:(sunil-demo2) ✗ python3 bin/demo.py pipeline --password=$PASSWORD --branch=sunil-demo2 http://sunil-f4j-publicsl-9adnwfg46s3s-1406902471.us-west-2.elb.amazonaws.com/ http://sunil-f4j-elasticl-1hx6ny1cqwidv-1482094555.us-west-2.elb.amazonaws.com/
[demo]: couldn't find marathon-lb running at 'http://sunil-f4j-publicsl-9adnwfg46s3s-1406902471.us-west-2.elb.amazonaws.com/'
```